Now it is possible to configure pac4j in the YAML

victornoel opened the pull request:
Now it is possible to specify global security filters (with `filters`), matchers, authorizers and the clients in the application's configuration.
For the clients, it is possible to specify the authenticator, and stuffs like that.

The pac4j configuration is now directly in the DW configuration, so no more need for `ConfigSingleton` to access it later.

Now, I need some feedbacks from you @evnm and @leleuj :)

@evnm I had to do a lot of tweaking with Jackson (using Mixins), is that the correct way to do it? Basically I wanted to be able to instantiate classes dynamically based on their name in the configuration.
There is some case where I would have preferred something better but I couldn't do it, for example, in the example, we have:
```yaml
pac4j:
  clients:
    clients:
      DirectBasicAuthClient:
        class: org.pac4j.http.client.direct.DirectBasicAuthClient
```
But it would have been nicer to be able ALSO to express it like this (if there is no other properties to set on the client):
```yaml
pac4j:
  clients:
    clients:
      DirectBasicAuthClient: org.pac4j.http.client.direct.DirectBasicAuthClient
```
The same with the other type of instantiable objects.

Also @evnm, I wanted to test only the configuration parsing, but I didn't find a way to do it without loading the whole application, not sure if it is possible? :)
And more generally, do you see something problematic?

@leleuj Did I forget some types that could be instantiated from the configuration? The list is in `Pac4jBundle` with the jackson Mixins as well as in the README :)
Do you see other things?

Thanks :)

leleuj commented:
In the *pac4j* project, we have a `pac4j-config` module which proposes to define your security configuration from properties.

Is your YAML solution something we could have in the core project and re-use in libraries or is it specific to Dropwizard?

victornoel commented:
@leleuj it is part of dropwizard but I don't know how easy it could be the same code used for both… I think it is possible, because Pac4jConfiguration as well as the Mixin themselves do not rely on dropwizard but only on Jackson, so if pac4j-config uses Jackson for parsing a configuration (YAML or other), then it could be used as-is, and then dropwizard-pac4j would reuse it.

I don't know how easy it would be to maintain in the long term, especially if dw and pac4j-config starts to use different versions of jackson :) But it could be done I think! @evnm what do you think on that point?

@leleuj as for the rest, any comments? :)

victornoel commented:
sorry, i didn't want to close it ^^

evnm commented on README.md:
```diff
@@ -39,50 +39,78 @@ Update the application's configuration class to expose accessor methods for `Pac
         }
     }
 
+Note that it is also possible to have `pac4jConfiguration` be null and in this case, the bundle will be disabled.
```
Please wrap text lines at 80 characters.

victornoel replied:
Ah yes, I set up editorconfig with a wrapping at 120 characters, I will change that.

evnm commented on src/main/java/org/pac4j/dropwizard/Pac4jBundle.java:
```diff
@@ -14,32 +21,39 @@
 /**
- * A {@link ConfiguredBundle} which installs a {@link SecurityFilter}
- * into a Dropwizard application's Jersey filter chain.
+ * A {@link ConfiguredBundle} which installs a {@link SecurityFilter} into a Dropwizard application's Jersey filter
```
Please wrap at 80 characters.

evnm commented on src/main/java/org/pac4j/dropwizard/Pac4jConfiguration.java:
```diff
@@ -1,7 +1,172 @@
+/**
+ * An interface defining getters for parameters required to configure a {@link SecurityFilter}.
+ */
+public class Pac4jConfiguration {
```
In other Dropwizard libraries, [classes like this](https://github.com/dropwizard/dropwizard/blob/master/dropwizard-db/src/main/java/io/dropwizard/db/DataSourceFactory.java) which encapsulate configuration options are termed "-Factory" rather than "-Configuration". In which case, this class would be `Pac4jFactory`, as before.

victornoel replied:
Ok, but I don't want it to "build" the configuration at the end, because I want users to be able to modify the config instance from the run() method of the `Application` (without having to use something like `ConfigSingleton` provided by pac4j). As that compliant with the Factory approach or is there an idiomatic way to achieve what I want?

evnm replied:
I think I'm missing something. Why exactly do you want users to be able to modify the config instance within `run` methods? The `Pac4jBundle#run` is invoked before the application's `run` method, so the Pac4j configuration options have already been applied by the time you get to application code.

More generally, I don't think modifying configuration at runtime is something we want to encourage. It's of course _possible_ (via config factories' setter methods), but it seems to me that it goes against the premise of out-of-process configuration which Dropwizard espouses (i.e. config by file or environment variables). Applications ingest configuration from some external source (which is easily editable by system administrators) rather than opaquely making decisions for themselves.

victornoel replied:
It's exactly what I want: that the user could continue configuring pac4j after its config has been populated by Pac4jBundle :)

I see your point theoretically for the non-need to modify it, but in practice it is not always true!
For example, with pac4j, I would like to be able to setup one of the authenticator with a connection to the database: how can I do that at another time than during the application.run method where I actually instantiate such a connection (via its factory's build method for example).
If you see another idiomatic way of doing that, I'm all for it, but even in the past when I only needed to access a Bundle's configuration, it was a problem because the configuration was built by a factory, so once created, the instance is not accessible anymore to anybody except the bundle! No?

victornoel replied:
The problem is that the db case is an example and I can see other cases where it would be needed.
I'm pretty sure there will be many cases where it will happen…
The alternative, is to stay in the dw's style and use a Factory and all BUT we would still use the ConfigSingleton provided by pac4j in order to give access to the Config for those who wants…
It would avoid making people do akward things where they manipulate dw configuration in the run method…

I will take a look at dropwizard-db in details, I don't use it personally, I build a mongodb client (I made my own MongoConfigurationFactory with a build method and all) in the run method, so it is not available before! But I haven't looked yet at dropwizard-db, so maybe there is something more elegant yes :)

evnm replied:
Bleh, this is tough. Take a look at dropwizard-db and let me know what you think.

I think a good strategy is for the path of least resistance to be that which is most idiomatically similar to other Dropwizard bundles. This may be, as you say, to target the simplest cases with a Factory class, leaving advanced use cases the ability to drop down to `ConfigSingleton`.

Going forward, we can gauge the level of usage of the factory and readjust our APIs if it ends up not suiting users' needs.

victornoel replied:
Ok, I will do that, use the idiomatic way and provide access to the config via `ConfigSingleton`.

evnm commented on src/main/java/org/pac4j/dropwizard/Pac4JMixins.java:
```diff
@@ -0,0 +1,63 @@
+        public abstract void setAuthorizationGenerators(AuthorizationGenerator... authorizationGenerators);
+
+        @JsonIgnore
+        public abstract void setAuthorizationGenerator(AuthorizationGenerator authorizationGenerator);
```
I'm not sure this method is necessary, given the varargs alternative above. The one-character difference in the method names could cause confusion.

victornoel replied:
It's not my choice, these are methods declared in the class to which I apply the mixin.
I wanted to avoid jackson trying to use this setter for setting an authorization generator, because if you look at the implementation, it's not a set, it is an add actually :)

evnm replied:
Gotcha.

victornoel commented:
Already committed the style fixes

victornoel commented:
By the way @evnm, what about simplifying the configuration of class to be instantiated?
(Note that the original description of the PR is not up to date since my last commit)

The use case is this:
```yaml
pac4j:
  clients:
    clients:
      - class: org.pac4j.http.client.direct.DirectBasicAuthClient
        authenticator:
          class: org.pac4j.http.credentials.authenticator.test.SimpleTestUsernamePasswordAuthenticator
```

versus

```yaml
pac4j:
  clients:
    clients:
      - class: org.pac4j.http.client.direct.DirectBasicAuthClient
        authenticator: org.pac4j.http.credentials.authenticator.test.SimpleTestUsernamePasswordAuthenticator
```

evnm commented:
Can we avoid the `clients` YAML node duplication?

It seems like we want to use the `JsonTypeInfo` annotations on properties of the outer `clients` node rather than on properties of intermediary mixin classes. The mixin approach is what is leading to the `clients` ->  `authenticator` -> `class` hoop-jumping.

victornoel commented:
The clients' clients duplication is there because it exists in pac4j, but I think that yes, we could do something about it, but then it means to move the other properties of the first clients into config… it's not a problem I guess!
As for the mixin @evnm, I'm not totally sure I understand, do you have some examples I could look at to get the idea? anyway, I will look at that this week-end either way :) Thanks!

evnm commented:
I think most of my commentary here is friction stemming from my lack of familiarity with pac4j. Sorry for this, and thank you for being so gracious with my semi-informed feedback. :)

On second though, I don't know if there's a way around `authenticator.class`. [Polymorphic deserialization in Jackson](http://wiki.fasterxml.com/JacksonPolymorphicDeserialization) is based on the use of specific marker properties which determine to which Java class adjacent member properties are fed. Thus for the `authenticator` node to be deserialized, it has to have some identifying property (i.e. `class`).

leleuj commented on src/main/java/org/pac4j/dropwizard/Pac4jBundle.java:
```diff
@@ -18,33 +26,51 @@
+        bootstrap.getObjectMapper().addMixIn(Authorizer.class,
+                Pac4JMixins.AuthorizerMixin.class);
+        bootstrap.getObjectMapper().addMixIn(Matcher.class,
+                Pac4JMixins.MatcherMixin.class);
```
So only these types will be available to create objects. In that case, the `RedirectActionBuilder` is missing as well as all types necessary for LDAP configuration (example: https://github.com/pac4j/pac4j/blob/master/pac4j-ldap/src/test/java/org/pac4j/ldap/test/tools/AuthenticatorGenerator.java#L19), MongoDB, SQL, but maybe it's too ambitious for a first version.

victornoel replied:
You are entirely right, it's missing, and it's not a problem at all to add it. Actually, I had it noted on a piece of paper but forgot about it ^^.
I would like to make the use case with a database work smoothly and be wholly configurable in the YAML file if possible. For now it isn't, mainly because of the interaction between the database as declared in the YAML and it's use in pac4j's configuration, also in the YAML file. This question is related to the one discussed with @evnm  just above: currently to fully setup one of the database authenticator, you need to do it outside of the config file.

leleuj commented:
This is where I realized I know little about Jackson ;-)

Just a comment about the defined classes which do not allow us to define LDAP support for example (see: http://www.pac4j.org/docs/authenticators/ldap.html).

If my understanding is correct...

victornoel commented:
And I don't know much about jackson, pac4j AND dropwizard :)

I will do another pass on all of this and see what comes out of it then!

victornoel commented:
I pushed some rework.

I changed back to the use of the Factory patter for the configuration, and `ConfigSingleton` to store pac4j's configuration.
I tried to make this code easy to extend to support other types and inject more mixins, not sure it's so good, but it's a start.
I rewrote the tests so that I only usr the `YamlConfigurationFactory`

@evnm I'm still not clear with respect to the database thing... I tried to look at how things were done with dropwizard-auth for example to support database-based authenticator, and I couldn't find any useful examples... Also, looking at dropwizard-db, I don't see how you manage to reuse the same datasource for every component that needs it in a dropwizard application: I feel like each bundle relying on it is simply building its own datasource by calling build() on the `DataSourceFactory`. To me it feels like, if there is multiple component accessing the database, that at one point, the db connection must be built and then passed to each of the components. In my application I do that directly in the `Application.run()` method, like [this](https://gitlab.com/victornoel/petals-cockpit/blob/dw-pac4j-config/backend/src/main/java/org/ow2/petals/cockpit/server/CockpitApplication.java#L82), [this](https://gitlab.com/victornoel/petals-cockpit/blob/dw-pac4j-config/backend/src/main/java/org/ow2/petals/cockpit/server/CockpitApplication.java#L84) and [this](https://gitlab.com/victornoel/petals-cockpit/blob/dw-pac4j-config/backend/src/main/java/org/ow2/petals/cockpit/server/CockpitApplication.java#L107).

evnm commented:
I may have misled you in sending you down the `dropwizard-db` path. `dropwizard-auth` doesn't provide any database glue code at all. Auth is handled by authenticator and authorizer classes, which may themselves use a DB interface (e.g. a DAO) to query a database to check credentials. I was thinking that any DB-dependent components of pac4j could be wired up to either an application's underlying `DataSourceFactory` or else a DAO constructed with `dropwizard-jdbi` or `dropwizard-hibernate`.

A couple question:

- Are there components _within pac4j itself_ that handle database interactions? Or are we talking about supporting application-defined authenticators that implement pac4j interfaces?
- What is the actual database interface that the relevant pac4j components require? For instance, does the "database-based authenticator" require a `javax.sql.DataSource` or could it use higher-level interfaces like a DAO powered by jDBI?

This could determine whether or not the pac4j components share the same underlying database connection pool used by the rest of the application. If multiple components of an application produce `DataSource`s with separate calls to `DataSourceFactory#build`, then you'll end up with independent DB clients. Not _necessarily_ a problem, but it doubles the number of connections that the database server must be able to handle.

victornoel commented:
@evnm basically, I would say we are in the case where "application-defined authenticators implement pac4j interfaces". Still, pac4j provides many implementations and most of them are actually not database oriented, they delegate authentication to things like openid, oauth, jwt, CAS, SAML, etc.
There is some database oriented implementations, but I don't think they are the heart of what pac4j is for, for example:
- SQL, taking a `javax.sql.DataSource` that is actually fed to a `org.skife.jdbi.v2.DBI` internally
- MongoDB, taking a `MongoClient`
- LDAP, taking a `org.ldaptive.auth.Authenticator`

Personally, I have my own implementation of the authenticator for MongoDB but using another library than the one used in pac4j-mongo.

leleuj commented:
LDAP, SQL and MongoDB supports are as "important" as CAS, SAML, OpenID Connect clients (for example). Though, clients address a wider scope while authenticators only validate credentials, they must be used in the HTTP clients. See: http://www.pac4j.org/docs/clients.html and http://www.pac4j.org/docs/authenticators.html, it should be clearer.

The extension capability is really better now. It seems good as much as I can see.

victornoel commented:
In you don't have any comments @evnm and @leleuj, I will merge that.
I updated the commit comment to explain what is important.

leleuj commented on src/main/java/org/pac4j/dropwizard/Pac4jMixins.java:
```diff
@@ -0,0 +1,45 @@
+    public abstract static class BaseClientMixin extends ClientMixin {
+
+        @JsonProperty
+        public abstract void setAuthorizationGenerators(
```
Maybe we'll need some cleaning the version 2.0 to make that easier.

leleuj commented:
+1
